### PR TITLE
FEAT: Ability to show video in notification as a gif

### PR DESCRIPTION
**Pull request by NickM-27**
## Summary

I have added the ability for a user to define `video` as part of an Android notification. This video will be downloaded and home assistant will attempt to parse frames from this video to show in a notification. The goal is to find 5 frames but it will work for any number of frames < 5 as well. 

## Screenshots

https://user-images.githubusercontent.com/14866235/156859680-5c76f67d-4e27-4ec5-a6ee-c6c96937ce91.mp4

## Link to pull request in Documentation repository

Documentation: home-assistant/companion.home-assistant#705

### CODE REVIEW

**Comment by dshokouhi** [comment]
This PR needs 1 additional PR to add the parameter to firebase as well, just need to add `video` here since its a new field

https://github.com/home-assistant/mobile-apps-fcm-push/blob/master/functions/android.js#L43-L52

**Comment by dshokouhi** [comment]
This PR made me think about https://github.com/home-assistant/android/issues/1270 I wonder if later on we can leverage the code here to fix it?

**Comment by NickM-27** [comment]
> This PR made me think about #1270 I wonder if later on we can leverage the code here to fix it?

yeah, sounds like there is a first party fix that works for API 31 but perhaps the `MediaMetaDataRetriever` can parse frames from gifs as well. I wouldn't mind investigating as part of this PR but perhaps that would belong on its own PR.

**Comment by dshokouhi** [comment]
> I wouldn't mind investigating as part of this PR but perhaps that would belong on its own PR.

lets do a separate PR for that to keep this one as small as possible :)

**Comment by NickM-27** [comment]
@dshokouhi Let me know what you think of the docs changes. One thing still in my head that I think may be worth adding in the notes is that videos less than 10 seconds long probably won't work very well since it may not find enough frames to show.

Also fcm PR: home-assistant/mobile-apps-fcm-push#705

**Comment by dshokouhi** [comment]
I just left a comment on the docs PR :) 



> One thing still in my head that I think may be worth adding in the notes is that videos less than 10 seconds long probably won't work very well since it may not find enough frames to show

I agree I think thats worth mentioning as a note specific to android as I dont think iOS has to worry about that. The only thing here that we want to make sure is that the service call can be the same for iOS and android :) Functionality doesn't need to match just that a user can put the 2 devices in a group and send the notification there.

File: app/src/main/java/io/homeassistant/companion/android/notifications/MessagingManager.kt [hunk 1]
```diff
@@ -1168,6 +1180,98 @@ class MessagingManager @Inject constructor(
...
+                            )
+                        }
+
+                        data[TITLE]?.let { rawTitle ->
+                            remoteViewFlipper.setTextViewText(R.id.title, rawTitle)
+                        }
+
+                        data[MESSAGE]?.let { rawMessage ->
+                            remoteViewFlipper.setTextViewText(R.id.info, rawMessage)
+                        }
```
Comment (JBassett) [inline app/src/main/java/io/homeassistant/companion/android/notifications/MessagingManager.kt:1210]: Doesn't the `handleText` function handle this?

Comment (NickM-27) [reply app/src/main/java/io/homeassistant/companion/android/notifications/MessagingManager.kt:1210]: No, when you set the style as `DecoreatedCustomViewStyle` you actually control the entire notification viewport so the title and text part from the builder is overwritten / not included and the custom view needs to include it itself.

Comment (NickM-27) [reply app/src/main/java/io/homeassistant/companion/android/notifications/MessagingManager.kt:1210]: To clarify on that last point, the docs here: https://developer.android.com/reference/android/app/Notification.DecoratedCustomViewStyle show how you can override specific parts of the notification. In this case I am overriding the bigContentView so the heads up view and the standard content view will use the `handleText` part but when the notification is expanded it needs to be handled by the custom view here.

**Comment by NickM-27** [comment]
@dshokouhi Is there anything else needed for this PR? Just curious what the timeline may be 😄

**Comment by dshokouhi** [comment]
Right now this PR is just waiting for a more official code review, depends on how much time the reviewer has.  Usually 1-2 weeks for a PR to be reviewed but could be longer, maybe faster if its fixing a critical bug :) There is a low PR count so its on the radar

**Comment by NickM-27** [comment]
> Right now this PR is just waiting for a more official code review, depends on how much time the reviewer has.  Usually 1-2 weeks for a PR to be reviewed but could be longer, maybe faster if its fixing a critical bug :) There is a low PR count so its on the radar

Thanks! Sounds good. By the way I looked into use the MediaMetaDataRetriever to parse gifs but unfortunately I don't believe that format is supported :/

I'll have to look and see if there are any other approaches that are lightweight. (I know Glide can do it but that's a big library for that problem alone)

**Review by JBassett — APPROVED** [review]

**Comment by conorlap** [comment]
I can't seem to get this working. I've installed [beta-2150-261ae31f](https://github.com/home-assistant/android/releases/tag/beta-2150-261ae31f) but no image/gif is appearing. Here's how I'm calling the notification:

```
service: notify.mobile_app_s20
data:
  message: Test
  data:
    video: "https://download.samplelib.com/mp4/sample-20s.mp4"

```
Am I doing something wrong?

**Pull request merged**